### PR TITLE
Refactor add constants

### PR DESCRIPTION
**Pull request by Jerakin**
Quite a big PR that pre-hashes strings for use as constants. I bet it will mess with our other branches, and I am sure I didn't find all instances where they should be replaced. Would prefer to get this merged in and then gradually change over the places missing. 

I opted to create a couple of files to divide the constant into separate parts.

`screens.` contains the screen names used with Monarch

`constants.` contains p5e specific constants

`messages.` contains messages that the app is using, both Defolds own and app specific

`platform.` contains constants that can be checked against to see what the current platform is.

### CODE REVIEW

File: utils/constants.lua [hunk 1]
```diff
@@ -0,0 +1,26 @@
+local M = {}
+
+M.ABILITY_LIST = {"STR", "DEX", "CON", "INT", "WIS", "CHA"}
+
+M.SR_LIST = {"1/8", "1/4", "1/2", "1", "2", "3", "4", "5", "6", "7","8", "9", "10", "11", "12", "13", "14", "15"}
+M.SR_TO_NUMBER = {["1/8"]=0.125, ["1/4"]=0.25, ["1/2"]=0.5, ["1"]=1, ["2"]=2, ["3"]=3, ["4"]=4, ["5"]=5, ["6"]=6, ["7"]=7,["8"]=8, ["9"]=9, ["10"]=10, ["11"]=11, ["12"]=12, ["13"]=13, ["14"]=14, ["15"]=15}
```
Comment (magroader) [inline utils/constants.lua:6]: Couldn't you set __index for this and avoid needing to define everything 1+?

File: utils/constants.lua [hunk 2]
```diff
@@ -0,0 +1,26 @@
+local M = {}
+
+M.ABILITY_LIST = {"STR", "DEX", "CON", "INT", "WIS", "CHA"}
+
+M.SR_LIST = {"1/8", "1/4", "1/2", "1", "2", "3", "4", "5", "6", "7","8", "9", "10", "11", "12", "13", "14", "15"}
```
Comment (magroader) [inline utils/constants.lua:5]: Are all of these SRs actually valid SRs from Pokemon that exist? This **could** be parsed during the Python script that scrapes all the pokemon.

File: utils/platform.lua [hunk 3]
```diff
@@ -0,0 +1,15 @@
+local M = {}
+
+local info = sys.get_sys_info()
+
+M.CURRENT = info.system_name
```
Comment (magroader) [inline utils/platform.lua:5]: I assume this is available immediately at boot and never changes?

File: utils/tracking_id.lua [hunk 4]
```diff
@@ -1,37 +1,40 @@
...
-	[hash("splash")] = 217,
-	[hash("storage")] = 218,
-	[hash("generate_pokemon")] = 219,
-	[hash("status_effects")] = 220,
-	[hash("import_pokemon")] = 221,
-	[hash("pokedex")] = 222,
-	[hash("settings")] = 223,
-	[hash("fakemon")] = 224,
-	[hash("moves_confirm")] = 225,
+	[screens.DEBUG] = 201,
```
Comment (magroader) [inline utils/tracking_id.lua:6]: I also changed this in my networking script. The merge shouldn't be bad, but maybe you'd want to do the same thing instead.

I gave tracking_id an init function and changed this [blah] = 20x into a flat array, then the init function loops through the array and assigns each a number. That way you don't need to increment a number every time, you just add the screen to the array.

**Comment by magroader** [comment]
This looks good to me. I didn't closely look at each switch from hardcoded string to constant, but I didn't notice any wrong. It's hard to tell if you missed any though :-)

File: utils/constants.lua (hunk 1 above)
Comment (Jerakin) [reply utils/constants.lua:6]: It's currently used to get the lowest and highest SR, in that case `__index` would do it. I could iterate over it and then add 1-15 to it like that.

File: utils/constants.lua [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+M.SR_TO_NUMBER = {["1/8"]=0.125, ["1/4"]=0.25, ["1/2"]=0.5, ["1"]=1, ["2"]=2, ["3"]=3, ["4"]=4, ["5"]=5, ["6"]=6, ["7"]=7,["8"]=8, ["9"]=9, ["10"]=10, ["11"]=11, ["12"]=12, ["13"]=13, ["14"]=14, ["15"]=15}
+
+
+M.NUMBER_TO_SR = {[0.125]="1/8", [0.25]="1/4", [0.5]="1/2"}
+
+-- Set a metatable that returns a string of the input number if it is missing
+setmetatable(M.NUMBER_TO_SR, {
+	__index = function(t, i)
+		if type(i) == "number" then
+			return tostring(i)
```
Comment (Jerakin) [inline utils/constants.lua:15]: This should probably return nil when it isn't a supported SR value. Something like this maybe

```lua
local sr = tostring(i)
return rawget(M.SR_TO_NUMBER, sr) ~= nil and sr or nil
```

Comment (Jerakin) [reply utils/constants.lua:15]: Using a metatable for this is maybe a bit.... superfluous too

File: utils/constants.lua (hunk 2 above)
Comment (Jerakin) [reply utils/constants.lua:5]: Could be! But they are all used so no need I think

File: utils/tracking_id.lua (hunk 4 above)
Comment (Jerakin) [reply utils/tracking_id.lua:6]: Yeah, probably a good idea. _Could_ make it a bit harder to verify which screen is which id though

File: utils/platform.lua (hunk 3 above)
Comment (Jerakin) [reply utils/platform.lua:5]: Yep

File: utils/constants.lua (hunk 5 above)
Comment (Jerakin) [reply utils/constants.lua]: Removed this: "explicit is better than implicit" after all

File: utils/tracking_id.lua (hunk 4 above)
Comment (Jerakin) [reply utils/tracking_id.lua:6]: Leaving it as is. Have decided to remove all the tracking except the error reporting anyway.

**Pull request merged**